### PR TITLE
Implement scan hook for validators info and all validators

### PR DESCRIPTION
**Pull request by Benzbeeb**

### CODE REVIEW

File: scan/src/rpc/JsonUtils.re [hunk 1]
```diff
@@ -3,4 +3,5 @@ module Decode = {
   let intstr = string |> map(int_of_string);
   let uamount = string |> map(float_of_string) |> map(v => v /. 1e6);
   let moment = string |> map(MomentRe.moment);
-};
+  let floatstr = string |> map(float_of_string)
+};
```
Comment (taobun) [inline scan/src/rpc/JsonUtils.re]: Endline!

File: scan/src/contexts/GlobalContext.re [hunk 2]
```diff
@@ -21,4 +21,4 @@ let make = (~children) => {
   };
 
   React.createElement(React.Context.provider(context), {"value": data, "children": children});
-};
+};
```
Comment (taobun) [inline scan/src/contexts/GlobalContext.re]: Endline!

File: scan/src/rpc/ValidatorHook.re [hunk 3]
```diff
@@ -18,13 +21,227 @@ module Validator = {
...
+    requestID: int,
+    txHash: string,
+    oracleScriptID: int,
+    oracleScriptName: string,
+    dataSourceIDList: list(int),
+    externalList: list(External.t),
+  };
+};
+
+let getValidator = (~limit=10, ~page=1, ~status="bonded", ()) => {
```
Comment (taobun) [inline scan/src/rpc/ValidatorHook.re]: getValidators

File: scan/src/contexts/GlobalContext.re [hunk 4]
```diff
@@ -11,7 +11,7 @@ let context = React.createContext(ContextHelper.default);
 let make = (~children) => {
   let financialOpt = PriceHook.get();
   let latestBlocksOpt = BlockHook.latest();
-  let validatorsOpt = ValidatorHook.get();
+  let validatorsOpt = ValidatorHook.getValidator();
```
Comment (taobun) [inline scan/src/contexts/GlobalContext.re]: getValidators

File: scan/src/rpc/ValidatorHook.re [hunk 5]
```diff
@@ -7,6 +7,9 @@ module Validator = {
     website: string,
     details: string,
     tokens: float,
+    commission: float,
+    uptime: float,
+    repoertRate: float,
```
Comment (taobun) [inline scan/src/rpc/ValidatorHook.re]: Merge ValidatorIndexInfo to this module

Comment (taobun) [inline scan/src/rpc/ValidatorHook.re]: Change report rate to Complete and Miss request count?

File: scan/src/rpc/ValidatorHook.re [hunk 6]
```diff
@@ -18,13 +21,227 @@ module Validator = {
...
+    rewardDestinationAddress: string,
+    votingPower: float,
+    commission: float,
+    bondedHeight: int,
+    website: string,
+    details: string,
+  };
+};
+
+module ValidatorIndexNodeSatus = {
```
Comment (taobun) [inline scan/src/rpc/ValidatorHook.re]: I think to add this data to Validator type

File: scan/src/rpc/ValidatorHook.re [hunk 7]
```diff
@@ -18,13 +21,227 @@ module Validator = {
...
+  type t = {
+    uptime: float,
+    avgResponseTime: int,
+  };
+};
+
+module ValidatorIndexRequestResponse = {
```
Comment (taobun) [inline scan/src/rpc/ValidatorHook.re]: Same with ValidatorIndexNodeSatus

**Review by taobun — CHANGES_REQUESTED** [review]

File: scan/src/contexts/GlobalContext.re [hunk 8]
```diff
@@ -11,7 +11,7 @@ let context = React.createContext(ContextHelper.default);
 let make = (~children) => {
   let financialOpt = PriceHook.get();
   let latestBlocksOpt = BlockHook.latest();
-  let validatorsOpt = ValidatorHook.get();
+  let validatorsOpt = ValidatorHook.getValidators();
```
Comment (evilpeach) [inline scan/src/contexts/GlobalContext.re]: `ValidatorHook.getList()`

File: scan/src/rpc/ValidatorHook.re [hunk 9]
```diff
@@ -1,30 +1,187 @@
...
+  type delegator_t = {
+    delegator: string,
+    sharePercentage: float,
+    amount: int,
+  };
+  type external_t = {
+    externalID: int,
+    externalValue: int,
+  };
+  type report_t = {
```
Comment (taobun) [inline scan/src/rpc/ValidatorHook.re:24]: Can reuse from `RequestHook.Report.t`

File: scan/src/rpc/ValidatorHook.re [hunk 10]
```diff
@@ -1,30 +1,187 @@
...
     details: string,
     tokens: float,
+    commission: float,
+    bondedHeight: int,
+    uptime: float,
+    completedRequestCount: int,
+    missedRequestCount: int,
+    indexNodeStatus: status_t,
+    indexRequestCountResponse: request_t,
+    proposedBlockList: list(proposed_block_t),
```
Comment (taobun) [inline scan/src/rpc/ValidatorHook.re]: proposedBlocks

File: scan/src/rpc/ValidatorHook.re [hunk 11]
```diff
@@ -1,30 +1,187 @@
...
+    delegatorList: list(delegator_t),
```
Comment (taobun) [inline scan/src/rpc/ValidatorHook.re]: ditto

File: scan/src/rpc/ValidatorHook.re [hunk 12]
```diff
@@ -1,30 +1,187 @@
...
     identity: string,
     website: string,
     details: string,
     tokens: float,
+    commission: float,
+    bondedHeight: int,
+    uptime: float,
+    completedRequestCount: int,
+    missedRequestCount: int,
+    indexNodeStatus: status_t,
```
Comment (taobun) [inline scan/src/rpc/ValidatorHook.re]: nodeStatus

File: scan/src/rpc/ValidatorHook.re [hunk 13]
```diff
@@ -1,30 +1,187 @@
 module Validator = {
+  type status_t = {
+    uptime: float,
+    avgResponseTime: int,
+  };
+  type request_t = {
```
Comment (taobun) [inline scan/src/rpc/ValidatorHook.re]: Remove this type if not used

File: scan/src/rpc/ValidatorHook.re [hunk 14]
```diff
@@ -1,30 +1,187 @@
 module Validator = {
+  type status_t = {
```
Comment (taobun) [inline scan/src/rpc/ValidatorHook.re]: node_status_t

File: scan/src/rpc/ValidatorHook.re [hunk 15]
```diff
@@ -1,30 +1,187 @@
...
+    timestamp: MomentRe.Moment.t,
+    blockHash: string,
+    txn: int,
+  };
+  type delegator_t = {
+    delegator: string,
+    sharePercentage: float,
+    amount: int,
+  };
+  type external_t = {
```
Comment (taobun) [inline scan/src/rpc/ValidatorHook.re]: Remove

File: scan/src/rpc/ValidatorHook.re [hunk 16]
```diff
@@ -1,30 +1,187 @@
 module Validator = {
+  type status_t = {
+    uptime: float,
+    avgResponseTime: int,
+  };
+  type request_t = {
+    completedRequestCount: int,
+    missedRequestCount: int,
+  };
+  type proposed_block_t = {
```
Comment (taobun) [inline scan/src/rpc/ValidatorHook.re]: Reuse `BlockHook.Block.t` instead

File: scan/src/rpc/ValidatorHook.re [hunk 17]
```diff
@@ -1,30 +1,187 @@
...
+module GlobalInfo = {
+  type t = {
+    allBondedAmount: int,
+    totalSupply: int,
+    inflationRate: float,
+    avgBlockTime: float,
+  };
+};
+
+module ValidatorStatus = {
```
Comment (taobun) [inline scan/src/rpc/ValidatorHook.re]: Make it to function

File: scan/src/rpc/ValidatorHook.re [hunk 18]
```diff
@@ -1,30 +1,187 @@
...
+  type t = {activeValidatorCount: int};
+
+  let decodeValidators = json =>
+    JsonUtils.Decode.{
+      activeValidatorCount: json |> field("result", list(_ => ())) |> Belt_List.length,
+    };
+};
+
+module ValidatorCount = {
```
Comment (taobun) [inline scan/src/rpc/ValidatorHook.re]: ditto

File: scan/src/rpc/ValidatorHook.re [hunk 19]
```diff
@@ -1,30 +1,187 @@
...
+    };
+};
+
+let getValidators = (~limit=10, ~page=1, ~status="bonded", ()) => {
+  let json = AxiosHooks.use({j|staking/validators?limit=$limit&page=$page&status=$status|j});
   json |> Belt.Option.map(_, Validator.decodeValidators);
 };
+
+let getGlobalInfo = _ => {
+  GlobalInfo.{
```
Comment (taobun) [inline scan/src/rpc/ValidatorHook.re]: . is necessary?

**Review by taobun — CHANGES_REQUESTED** [review]

File: scan/src/rpc/ValidatorHook.re (hunk 9 above)
Comment (Benzbeeb) [reply scan/src/rpc/ValidatorHook.re:24]: report_t has `requestID` so it isn't equal to RequestHook.Report.t

**Review by evilpeach — CHANGES_REQUESTED** [review]

File: scan/src/rpc/ValidatorHook.re [hunk 20]
```diff
@@ -1,30 +1,169 @@
...
+    totalSupply: 10849023,
+    inflationRate: 12.45,
+    avgBlockTime: 2.59,
+  };
+};
+
+let decodeValidators = json =>
+  JsonUtils.Decode.(json |> field("result", list(_ => ())) |> Belt_List.length);
+
+let getValidatorStatus = (~status="bonded", ()) => {
```
Comment (sorawit) [inline scan/src/rpc/ValidatorHook.re]: Why does this return an int?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: scan/src/rpc/ValidatorHook.re [hunk 21]
```diff
@@ -1,30 +1,170 @@
...
+  let json = AxiosHooks.use({j|staking/validators?status=$status|j});
+  let len = json |> Belt.Option.map(_, decodeValidators) |> Belt_Option.getWithDefault(_, 0);
+  len > 0;
+};
+
+let getValidatorCount = _ => {
+  let json = AxiosHooks.use({j|staking/validators|j});
+  json |> Belt.Option.map(_, decodeValidators);
+};
```
Comment (evilpeach) [inline scan/src/rpc/ValidatorHook.re]: The default status of `staking/validators` is `bonded`. So, we should combine `getValidatorStatus` and `getValidatorCount` together.

File: scan/src/rpc/ValidatorHook.re [hunk 22]
```diff
@@ -1,30 +1,170 @@
...
+    uptime: float,
+    completedRequestCount: int,
+    missedRequestCount: int,
+    nodeStatus: node_status_t,
+    proposedBlocks: list(block_t),
+    delegators: list(delegator_t),
+    reports: list(report_t),
   };
 
   let decodeValidator = json =>
```
Comment (evilpeach) [inline scan/src/rpc/ValidatorHook.re]: change the function name to `decodeResult`

File: scan/src/rpc/ValidatorHook.re [hunk 23]
```diff
@@ -1,30 +1,170 @@
...
+          oracleScriptID: 112,
+          oracleScriptName: "US powerball",
+          dataSourceIDList: [3, 4],
+          externalIDs: [3, 110],
+          externalValues: [4, 32],
+        },
+      ],
     };
 
   let decodeValidators = json =>
```
Comment (evilpeach) [inline scan/src/rpc/ValidatorHook.re]: change to decodeList

File: scan/src/rpc/ValidatorHook.re [hunk 24]
```diff
@@ -1,30 +1,170 @@
...
+          hash:
+            Hash.fromBase64("a469056894d91a4987ef2a07f16fc7bbae04f8166b43aa168e71e48b71cac651"),
+          oracleScriptID: 112,
+          oracleScriptName: "US powerball",
+          dataSourceIDList: [3, 4],
+          externalIDs: [3, 110],
+          externalValues: [4, 32],
+        },
+      ],
     };
```
Comment (evilpeach) [inline scan/src/rpc/ValidatorHook.re:134]: add decode for the single validator

**Review by evilpeach — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

**Review by evilpeach — APPROVED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**